### PR TITLE
Cookbook guide for ssd1306 OLED

### PR DESCRIPTION
**Pull request by erazor666**
## Description:
Cookbook guide for the ssd1306 OLED display with some examples and code

**Related issue (if applicable):** fixes <link to issue>
N/A
**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#<esphome PR number goes here>
N/A
**Pull request in [esphome-core](https://github.com/esphome/esphome-core) with C++ framework changes (if applicable):** esphome/esphome-core#<esphome-core PR number goes here>
N/A
## Checklist:

  - [ ] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.

### CODE REVIEW

**Comment by ahd71** [comment]
Hi. Looks like we hade the same idea (see pull request 173) :-)

Your guide and exempel were covering slighlty more tho (Images, i2c scan for example) so perfectly fine if admins publish yours instead of mine.

Well done!

**Comment by erazor666** [comment]
Oh... I see, what a strange coinicidence. Well, i'm guessing there is room for everyone here :) I started working on this a couple of days ago after working with displays some, and wishing there was a small guide for it, had to do some reading to get to where i wanted :) i guess its up to Otto what he wants.

File: cookbook/oled.rst [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+===========================
+
+| All the info in this guide can be found in the documentation :doc:`ssd1306_i2c </components/display/ssd1306_i2c>` and :doc:`Display </components/display/>`. 
+| In this article i will show some practical examples on how to use it, and maybe it can help beginners getting started using this display. The documentation will have more in depth info on all the options available.
+| These displays are pretty cheap on `AliExpress <https://www.aliexpress.com/item/10pcs-0-96-yellow-blue-0-96-inch-OLED-module-New-128X64-OLED-LCD-LED-Display/32638669209.html>`__ and they are quite bright.
+| There is also a 1.3" version with same resolution that should work with the same config as this one, however i have not tested personally yet.
+|
+| The display used in the pictures is a dual color display. X 0-127 through Y 0-15 are yellow pixels, the rest are blue. There are different versions of the display.
+
+| The display itself is a 0.96” OLED unit that connects to your ESP via `i2c`, you will need to add `i2c` configuration to your node.
```
Comment (OttoWinter) [inline cookbook/oled.rst]: \`i2c\` -> \`\`i2c\`\` - literals are with double backticks in rst.

File: cookbook/oled.rst [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+.. code-block:: yaml
+
+    i2c:
+      sda: D5
+      scl: D6
+      scan: true
+
+Set ``scan: true`` if you don’t know the address of your display and check the log for the node:
+
+| **[11:54:53][C][i2c:043]: I2C Bus:**
```
Comment (OttoWinter) [inline cookbook/oled.rst]: Please put this in a ``code-block``

```rst
.. code-block::

    [11:54:53][C][i2c:043]: I2C Bus:
    ...
```

File: cookbook/oled.rst [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+| **[11:54:53][C][i2c:044]:   SDA Pin: GPIO14**
+| **[11:54:53][C][i2c:045]:   SCL Pin: GPIO12**
+| **[11:54:53][C][i2c:046]:   Frequency: 50000 Hz**
+| **[11:54:53][I][i2c:048]: Scanning i2c bus for active devices...**
+| **[11:54:53][I][i2c:055]: Found i2c device at address 0x3C**
+|
+| Take note of the address and set ``scan: false`` again.
+| **(The board used in this example is a Wemos D1 mini, so adapt pins to the board you are using)**
```
Comment (OttoWinter) [inline cookbook/oled.rst]: Don't emphasize the whole sentence, if at all (looks overly aggressive in docs).
Maybe just the last part after `so`

File: cookbook/oled.rst [hunk 4]
```diff
@@ -0,0 +1,254 @@
+A guide to using SSD1306 OLED display
+===========================
```
Comment (OttoWinter) [inline cookbook/oled.rst]: The length of the header line must match the title length, so add a couple of `=` here (until it matches the text above). Also for other items here.

File: cookbook/oled.rst [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+| **[11:54:53][I][i2c:055]: Found i2c device at address 0x3C**
+|
+| Take note of the address and set ``scan: false`` again.
+| **(The board used in this example is a Wemos D1 mini, so adapt pins to the board you are using)**
+
+Node config:
+*********
+
+| The configuration for your node should look something like this: 
+| (your basic node configuration, eg network and such is not discussed in this guide):
```
Comment (OttoWinter) [inline cookbook/oled.rst]:
```suggestion
| (your basic node configuration, e.g. network and such is not discussed in this guide):
```

File: cookbook/oled.rst [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+          // thermometer image, thermo.png
+          it.image(0, 26, id(thermo));
+
+
+In detail:
+_________
+
+Sensors:
+*******
+.. code-block:: yaml
```
Comment (OttoWinter) [inline cookbook/oled.rst:117]: Add an empty line after heading

File: cookbook/oled.rst [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        id: wifi
+      - platform: homeassistant
+        entity_id: sensor.outside_temp
+        id: temp
+      - platform: homeassistant
+        entity_id: sensor.outside_rh
+        id: rh
+
+| This will set up your sensors, first one is internal from the node, the two next are imported from Home Assistant.
+| Don't give names to sensors you wish to hide from HA (no point duplicating sensors).
```
Comment (OttoWinter) [inline cookbook/oled.rst]: Don't use contractions like HA in the docs - in instant messaging it is useful to save space, but not here.

File: cookbook/oled.rst [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+*******
+.. code-block:: yaml
+
+    font:
+      - file: "roboto.ttf"
+        id: robo12
+        size: 12
+
+| Embeds the roboto font into your sketch, for more sizes make more of the same font with different sizes.
+| You can have multiple fonts and sizes in your sketch. The font used in this example is `Roboto Condensed light <http://allfont.net/download/roboto-condensed-light/>`__
```
Comment (OttoWinter) [inline cookbook/oled.rst:157]: I cannot find a ttf file download link on that website.

Let's use the website from the maker of this font: https://fonts.google.com/?selection.family=Roboto:900&thickness=7&query=roboto&category=Sans+Serif
Or even better the direct download link: https://fonts.google.com/download?family=Roboto

File: cookbook/oled.rst [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+      - file: "water.png"
+        id: water
+        resize: 20x20
+
+| Icons used in this sketch are imported from `MaterialDesigns <https://materialdesignicons.com/>`__ and are the same as are used in Home Assistant.
+| The 3 icons used are ``mdi:signal`` ``mdi:water-percent`` and ``mdi:thermometer``
+| You will need to download the icons from `MaterialDesigns <https://materialdesignicons.com/>`__ and put them in the same folder as your node config .yaml file is stored.
+
+.. note::
+     The icons from MaterialDesigns can come with a transparent background so you may have to fire up your favorite image editor and give them a white background if they aren't displayed correctly
```
Comment (OttoWinter) [inline cookbook/oled.rst:182]: Insert a blank line after `.. note::`.

Do transparent fonts not work with esphome? I thought they did...

File: cookbook/oled.rst [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+.. figure:: images/oled-topbar.png
+    :align: center
+    :width: 80.0%
+
+.. figure:: images/oled-temphum.png
+    :align: center
+    :width: 80.0%
+
+References
```
Comment (OttoWinter) [inline cookbook/oled.rst]:
```suggestion
See Also
```

File: cookbook/oled.rst [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+______________
+:doc:`Display </components/display>`
+:doc:`ssd1306_i2c </components/display/ssd1306_i2c>`
+:doc:`Time </components/time#strftime>`
+:doc:`Images </components/display/#images>`
```
Comment (OttoWinter) [inline cookbook/oled.rst]: Make this into a list please, with `- ` (as in other pages)

File: cookbook/oled.rst [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+.. figure:: images/oled-temphum.png
+    :align: center
+    :width: 80.0%
+
+References
+______________
+:doc:`Display </components/display>`
+:doc:`ssd1306_i2c </components/display/ssd1306_i2c>`
+:doc:`Time </components/time#strftime>`
```
Comment (OttoWinter) [inline cookbook/oled.rst]: Do not append `#strftime` here, it does not work

File: cookbook/oled.rst (hunk 11 above)
Comment (OttoWinter) [inline cookbook/oled.rst]: Add a disqus directive below this and a `ghedit` one too (see other docs pages)

File: index.rst [hunk 13]
```diff
@@ -297,6 +297,7 @@ Cookbook
     BME280 Environment, cookbook/bme280_environment, bme280.jpg
     Sonoff Fishpond Pump, cookbook/sonoff-fishpond-pump, cookbook-sonoff-fishpond-pump.jpg
     H801 LED Controller, cookbook/h801, h801.jpg
+    SSD1306 I2C Guide, cookbook/oled, oled.png
```
Comment (OttoWinter) [inline index.rst]: The oled.png image is too big, 158kb is too much for the main index - see contributing guide for how to shrink it down.

File: cookbook/oled.rst [hunk 14]
```diff
@@ -0,0 +1,254 @@
+A guide to using SSD1306 OLED display
+===========================
+
+| All the info in this guide can be found in the documentation :doc:`ssd1306_i2c </components/display/ssd1306_i2c>` and :doc:`Display </components/display/>`. 
+| In this article i will show some practical examples on how to use it, and maybe it can help beginners getting started using this display. The documentation will have more in depth info on all the options available.
```
Comment (OttoWinter) [inline cookbook/oled.rst]: `i` -> `I`
Or even better, no I: `This article will show some ...`

**Review by OttoWinter — COMMENTED** [review]
Ok, I think we can add both docs - we can always clean things up again afterwards. Plus this one contains nice images :)

@erazor666 Please remove all `| ` at the beginning of lines. I know it is used to break lines directly, but that ends up looking horrible on mobile devices. So please remove `| ` from the beginning of all lines. (look at other docs pages for how to lay things out, and contributing guide; plus netlify build preview in PR checks)

File: cookbook/oled.rst (hunk 3 above)
Comment (erazor666) [reply cookbook/oled.rst]: The document system doesn't seem to "respect" line breaks without the | its a bit limiting in that sense :-p

File: cookbook/oled.rst (hunk 8 above)
Comment (erazor666) [reply cookbook/oled.rst:157]: http://allfont.net/download/roboto-condensed-light/#download_font_block

It's there in a zip file, the condensed version doesn't seem to be on the google page. I can add both links i guess. It's not really a big deal what font you use, i was looking for one that looked somewhat ok on a small screen.

File: cookbook/oled.rst (hunk 9 above)
Comment (erazor666) [reply cookbook/oled.rst:182]: There is something 'strange' with the Mdi icons sometimes. I encountered it, would just show a colored square that matched the resize you have selected in your sketch so i did some ms paint messing about with them and then they came to life. Not sure if its Esphome or material designs that is the culprit.

I have not tested a whole lot of fonts, so for transparent fonts i don't know.

**Comment by erazor666** [comment]
And please forgive my stumbling around github, first ever project I take part in. I will get better at it.
After some 25 years on various instant messaging apps my writing has suffered somewhat.

I am planing on making some guides when I make something that other people could benefit from, the document format is kind of 'furstrating' to work with (or my methods). I've resorted to forking the repo to my own github, then i write the code in visual studio code. In order to have a look at the document I have to upload it to my own fork and inspect it there. Some of the directives like note, :doc: etc aren't represented correctly there however. I think I will get more comfortable with the code requirements as a I get more experience with it though. 

If you have suggestions for a better 'work environment' I'm all ears 😃

File: cookbook/oled.rst (hunk 8 above)
Comment (erazor666) [reply cookbook/oled.rst:157]: If you have no objections i'll leave it as it is, since the page has a download link, can always be changed later

**Review by OttoWinter — COMMENTED** [review]
- This PR contains commits that are not part of this PR (see index.rst, there are changes that are not part of this PR). If you create a fork, you should create a new branch for each feature.

```bash
git remote add upstream https://github.com/esphome/esphome-docs
git fetch upstream
git checkout upstream/current
git checkout -b ssd1306-oled
# Re-Add changes
git add .
git commit -m "Cookbook guide for SSD1306 OLED"
# Submit PR
```

- See travis output, there are a bunch of lint issues with this PR (newline after heading etc)
- Headings should be "Title Case", so `Clock(time):` -> `Clock (Time)`, `The display config:` -> `The Display Config`

**Comment by erazor666** [comment]
Remaking this PR with updated docs

**Pull request closed without merging**